domain/domain.go

```diff
@@ -4,32 +4,169 @@ import (
 	"github.com/kaspanet/kaspad/domain/consensus"
 	"github.com/kaspanet/kaspad/domain/consensus/model/externalapi"
 	"github.com/kaspanet/kaspad/domain/miningmanager"
+	"github.com/kaspanet/kaspad/domain/prefixmanager"
 	infrastructuredatabase "github.com/kaspanet/kaspad/infrastructure/db/database"
+	"github.com/pkg/errors"
+	"sync/atomic"
+	"unsafe"
 )
 
 // Domain provides a reference to the domain's external aps
 type Domain interface {
 	MiningManager() miningmanager.MiningManager
 	Consensus() externalapi.Consensus
+	TemporaryConsensus() externalapi.Consensus
+	CreateTemporaryConsensus() error
+	CommitTemporaryConsensus() error
 }
-
 type domain struct {
-	miningManager miningmanager.MiningManager
-	consensus     externalapi.Consensus
+	miningManager   miningmanager.MiningManager
+	consensus       *externalapi.Consensus
+	tempConsensus   *externalapi.Consensus // We assume there's no concurrent access to tempConsensus
+	consensusConfig *consensus.Config
+	db              infrastructuredatabase.Database
+}
+
+func (d *domain) Consensus() externalapi.Consensus {
+	return *d.consensus
 }
 
-func (d domain) Consensus() externalapi.Consensus {
-	return d.consensus
+func (d *domain) TemporaryConsensus() externalapi.Consensus {
+	return *d.tempConsensus
 }
 
-func (d domain) MiningManager() miningmanager.MiningManager {
+func (d *domain) MiningManager() miningmanager.MiningManager {
 	return d.miningManager
 }
 
+func (d *domain) CreateTemporaryConsensus() error {
+	_, hasInactivePrefix, err := prefixmanager.InactivePrefix(d.db)
+	if err != nil {
+		return err
+	}
+
+	if hasInactivePrefix {
+		return errors.Errorf("cannot have more than one inactive prefix")
+	}
+
+	activePrefix, exists, err := prefixmanager.ActivePrefix(d.db)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		return errors.Errorf("cannot create a temporary consensus when there's " +
+			"no active consensus")
+	}
+
+	var inactivePrefix byte = 0
+	if activePrefix == 0 {
+		inactivePrefix = 1
+	}
+
+	err = prefixmanager.SetPrefixAsInactive(d.db, inactivePrefix)
+	if err != nil {
+		return err
+	}
+
+	consensusFactory := consensus.NewFactory()
```

Comment:
Not sure this is very important, but maybe we can have one factory for every time?

Reply:
IMO keeping the factory as part of domain state adds unnecessary complexity with no real gain